enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,701 @@
+* Increases support complexity due to integration with the a user's provider network.
+* Increases network control plane load on nodes, as they have multiple peer connections and are having to maintain
+a dynamic routing protocol.
+
+### Implementation Details/Notes/Constraints
+
+#### BGP Peering
+
+FRR-K8S shall support both internal BGP (iBGP) and external BGP (eBGP). With iBGP, a full mesh topology is required. This
+can be cumbersome and have issues with scaling in large clusters. In the future, FRR-K8S shall also support designating one
+or more nodes as route reflectors (https://issues.redhat.com/browse/CNF-10719). With route reflectors, nodes only have
+to connect to the route reflectors and do not have to maintain a full mesh with all other nodes in the cluster. Therefore,
+support for iBGP in this enhancement will be restricted to full mesh only.
+
+#### FRR-K8S Integration
+
+As previously mentioned frr-k8s will be used in order to deploy and manage FRR configuration. Support will be added
+where necessary for gaps in frr-k8s API, but in the meantime frr-k8s supports also issuing raw FRR configuration. Every
+attempt should be made to leverage the FRR-K8S API, and raw configuration should only be used as a last resort. A user
+will need the ability to:
+
+* Define BGP peers, ASN, authentication, along with other common BGP configurations like communities, route filtering, etc.
+* Configure BFD
+
+FRR-K8S will serve as the main API for user configuration of BGP via FRR-K8S CRs. OVN-Kubernetes will watch for these
+CRs, and once created, OVN-Kubernetes will start monitoring the kernel routing table via netlink for routes installed
+by FRR. When FRR is using Zebra, the routes can be identified via a proto "bgp" label:
+
+```
+123.123.123.0/24 nhid 47 via 172.18.0.5 dev breth0 proto bgp metric 20
+```
+
+When these routes are seen on the node, ovnkube-controller will update the routes in NBDB for any GW routers on the
+node, filtering the route for the GW router based on if the interface on the route matches the interface in OVS. This
+will happen irrespective of gateway mode, as some features in local gateway mode rely on the GRs for sending traffic
+(Egress IP, MEG). Note, this configuration will only happen on nodes selected by the FRR-K8S CRs.
+
+#### API Changes
+
+##### OpenShift API
+
+The OpenShift API will be modified in order to allow CNO to deploy FRR and FRR-K8S CRDs:
+
+```golang
+// NetworkSpec is the top-level network configuration object.
+type NetworkSpec struct {
+    // ...
+    // deployFRR specifies whether or not the Free Range Routing (FRR) stack
+    // along with FRR-K8S API should be deployed by the operator.
+    // FRR is required for enabling for using any OpenShift features that require dynamic
+    // routing. This includes BGP support for OVN-Kubernetes and MetalLB.
+    // +optional
+    DeployFRR *bool `json:"deployFRR,omitempty"`
```

Comment:
@jcaamano @fedepaol we need to discuss this a bit. Do we want to allow FRR to only be deployed on a subset of nodes? How does it work with metallb operator today?

Reply:
In metallb today you can specify a node selector for the nodes you want to run the speakers on.
Currently the operator leverages the same node selector to deploy frr-k8s.

I'd say, a conservative approach is to deploy frr-k8s on all the nodes and still use that node selector for the speakers. Each speaker configures its own node so the will of the user to pick a subset of the nodes will be respected.

I think we need to expand the "BGP support for ovnk" bit behind this flag, as if it means reconfigure & restart ovnk, it's a bit harsh for the scenario where the user doesn't enable it from the start but then deploys metallb.

Reply:
Also also, for the original question, if the need arises, we can always add a node selector later, should be fine from an API point of view.
One last note, we'll have to deploy it on the master nodes too.

Reply:
If we should add a nodeSelector to manage which nodes FRR are deployed to, then why wouldn't we also do the same thing for FRR-K8S? Seems like an easy thing to do.

For OVNK I see 2 options:
1. Always register the RouteAdvertisements CRD and FRR-K8S CRDs in CNO, regardless of FRR flag. A user can attempt to create a CR, but it wont do anything because OVNK will complain there is no *valid* (maybe we can check FRR object status) FRR-K8S configuration. OVNK doesn't have to restart as its always listing and watching the resources.
2. Only register the CRDs when the deployFRR flag is set to true. In this case, OVNK will need to be rolled out if someone enables the flag post install so it can start watchers for the CRDs.

Reply:
I don't think 1 vs 2 is an issue. Actually what we do today for any other optional feature is a mix of both: always register the CRDs but also hide the functionality behind a OVNK cli flag for which we would need a roll out to set it. I don't see a reason to do it any differently. Also, we now have an additional reason to have a OVNK cli flag: the 36 month lifecycle for which we decided to gate any new feature behind a cli flag.

The concern here was to have 1 vs 2 api flags (https://github.com/openshift/enhancements/pull/1636#discussion_r1653015726), and I guess we will settle with 1 for now unless we need something else to resolve the upgrade scenario (https://github.com/openshift/enhancements/pull/1636#discussion_r1655585055) but I don't see how would that help.

`DeployFRR` seems to be an implementation detail. What about `EnableBGP`?

Reply:
> I don't think 1 vs 2 is an issue. Actually what we do today for any other optional feature is a mix of both: always register the CRDs but also hide the functionality behind a OVNK cli flag for which we would need a roll out to set it. I don't see a reason to do it any differently. Also, we now have an additional reason to have a OVNK cli flag: the 36 month lifecycle for which we decided to gate any new feature behind a cli flag.
> 

I would've said that not having the ifs around the codebase, and not having CNO to reconcile the flag would've resulted in a lower complexity, but I won't question it as you folks know it better than me.

I'd insist on the fact that seeing ovnk rollout if you deploy metallb is not optimal and we should avoid it if possible. When instructed to deploy (via the metallb crd), the metallb operator should set a flag somewhere and wait for frr-k8s to pop up.

> The concern here was to have 1 vs 2 api flags ([#1636 (comment)](https://github.com/openshift/enhancements/pull/1636#discussion_r1653015726)), and I guess we will settle with 1 for now unless we need something else to resolve the upgrade scenario ([#1636 (comment)](https://github.com/openshift/enhancements/pull/1636#discussion_r1655585055)) but I don't see how would that help.
> 
> `DeployFRR` seems to be an implementation detail. What about `EnableBGP`?

This might be required for EVPN too though.

Reply:
> I'd insist on the fact that seeing ovnk rollout if you deploy metallb is not optimal and we should avoid it if possible. 

I don't understand what's the source of the concern. Handling configuration changes and making the necessary rollouts is one of the main use cases of CNO. If the concern is that this happens without user interaction, perhaps the questionable decision here is having MetalLB triggering the CNO flag itself. Perhaps it should just check it and if not set, hold and report a relevant status condition.

While we might find a way to work around it now, I wouldn't count on upholding CNO from fulfilling one of its main use cases in the future, meaning, it likely in the future that we will have a reason to rollout even if we don't have it today.

Reply:
> > I'd insist on the fact that seeing ovnk rollout if you deploy metallb is not optimal and we should avoid it if possible.
> 
> I don't understand what's the source of the concern. Handling configuration changes and making the necessary rollouts is one of the main use cases of CNO. If the concern is that this happens without user interaction, perhaps the questionable decision here is having MetalLB triggering the CNO flag itself. Perhaps it should just check it and if not set, hold and report a relevant status condition.
> 

The concern is about having a user with a functioning cluster, they enable metallb which causes OVNK to roll out again (and with that, I guess, some downtime? If not, ignore what I just wrote).

> While we might find a way to work around it now, I wouldn't count on upholding CNO from fulfilling one of its main use cases in the future, meaning, it likely in the future that we will have a reason to rollout even if we don't have it today.

It's not a workaround. FRR-K8s won't serve only ovnk needs. It's now being deployed by CNO because it's handy to have it available at the same time OVN-K start, so pod 2 pod cidrs will be announced via BGP, but it's something that it's orthogonal to BGP - OVNK. There are users (including those using MetalLB) which might only care about the FRR-K8s api (or metallb) but not BGP-OVNK.

Reply:
> The concern is about having a user with a functioning cluster, they enable metallb which causes OVNK to roll out again (and with that, I guess, some downtime? If not, ignore what I just wrote).

Yes, there is a downtime and the user should be informed. But this is no different than most other configuration changes for OVN-K.

> It's not a workaround. FRR-K8s won't serve only ovnk needs. It's now being deployed by CNO because it's handy to have it available at the same time OVN-K start, so pod 2 pod cidrs will be announced via BGP, but it's something that it's orthogonal to BGP - OVNK. There are users (including those using MetalLB) which might only care about the FRR-K8s api (or metallb) but not BGP-OVNK.

If it is orthogonal and we are concerned about the downtime, then we might be better served by two separate API flags. Doing something else is what would feel somewhat like a workaround to me: listing and watching resources always (even if not doing much) takes up resources anyway and we don't have absolute control since a user could still deploy things manually; not too concerning but it wouldn't be my preferred option.

I wasn't too concerned about the downtime itself and I suspect @trozet might haven't been either.

Reply:
> I think the difference here is, the user only expects to deploy metallb here (as he was doing before) and he gets also a downtime (OR, we ask him to set two flags to deploy metallb and still expect the downtime, which is not optimal but at least expected).

OK so we would consider the downtime or requiring manually setting an extra flag a breaking change for users of the existing workflow. I can get aligned behind this idea.

I would preferably opt for two flags vs having the functionality always active OVN-K side. I don't see downsides for the two API flags, other than not all combinations of their values are valid but that can be properly validated.

Reply:
> > `DeployFRR` seems to be an implementation detail. What about `EnableBGP`?
> 
> This might be required for EVPN too though.

`EnableAdvancedRouting` maybe? Since I guess all FRR capabilities could the theoretically be available and used?

Reply:
Or maybe even `EnableFRR`, I guess the detail would be relevant anywary since we wouldn't break users already using it in the future. `DeployFRR` not too different but it does now sound too `imperative` to me.

Reply:
Unless we plan to enable additional routing components other than FRR under the same flag in the future

Reply:
To summarize this, we decided:
1. no nodeSelector required. Just deploy on every node for now. This may change int he future so we should make the config an extendable struct.
2. We will use 2 API flags. 1 for deploying FRR, and one to control support for BGP integration with OVNK.

Will update the doc.